File: src/main/java/leetcode/hot100/LeetCode215.java
package leetcode.hot100;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/12 23:33
 */
public class LeetCode215 {


    //https://leetcode.cn/problems/kth-largest-element-in-an-array/solution/shu-zu-zhong-de-di-kge-zui-da-yuan-su-by-leetcode-/
    public int findKthLargest(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int index = part(nums, left, right);
            if (index + 1 == k) {
                return nums[index];
            } else if (index + 1 < k) {
                left = index + 1;
            } else {
                right = index - 1;
            }
        }
        return -1;
    }

    private int part(int[] nums, int i, int j) {
        int h = nums[i];
        while (i < j) {
            while (nums[j] <= h && i < j) {
                j--;
            }
            nums[i] = nums[j];
            while (nums[i] >= h && i < j) {
                i++;
            }
            nums[j] = nums[i];
        }
        nums[i] = h;
        return i;
    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode215.go
package hot100

func findKthLargest(nums []int, k int) int {
	if len(nums) == 0 {
		return 0
	}
	if len(nums) < k {
		return -1
	}
	left, right := 0, len(nums)-1
	for left <= right {
		index := findKthLargestHelper(nums, left, right)
		if index+1 == k {
			return nums[index]
		} else if index+1 < k {
			left = index + 1
		} else {
			right = index - 1
		}
	}
	return -1
}

func findKthLargestHelper(nums []int, left, right int) int {
	pi := nums[left]
	for left < right {
		// 搞清楚顺序
		for left < right && nums[right] <= pi {
			right--
		}
		nums[left] = nums[right]
		for left < right && nums[left] >= pi {
			left++
		}
		nums[right] = nums[left]
	}
	nums[left] = pi
	return left
}
